### PR TITLE
Add contribution guidelines, pre-commit and releasenotes

### PR DESCRIPTION
**Pull request by taalexander**
This PR adds support for contribution guidelines, pre-commit and releasenotes to the qe-compiler.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+    details.
+```
+
+### Generating the release notes
+
+After release notes have been added, you can use reno to see what the full output
+of the release notes is. In general the output from reno that we'll get is a rst
+(ReStructuredText) file that can be compiled by
+[sphinx](https://www.sphinx-doc.org/en/master/). To generate the rst file you
+use the ``reno report`` command. If you want to generate the full qw-compiler release
```
Comment (blakejohnson) [inline CONTRIBUTING.md:161]:
```suggestion
use the ``reno report`` command. If you want to generate the full qe-compiler release
```

Comment (taalexander) [reply CONTRIBUTING.md:161]: Fixed in 049906b.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+changes over a release cycle we require that all user facing changes include
+documentation at the same time as the code. To accomplish this we use the
+[reno](https://docs.openstack.org/reno/latest/) tool which enables a git based
+workflow for writing and compiling release notes.
+
+### Adding a new release note
+
+Making a new release note is quite straightforward.
+
+- Ensure that you have renoinstalled with:
```
Comment (blakejohnson) [inline CONTRIBUTING.md:91]:
```suggestion
- Ensure that you have reno installed with:
```

Comment (taalexander) [reply CONTRIBUTING.md]: Fixed in 7af8d5a

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+- Check for no merge conflict lines accidentally being staged
+- Clang format
+- Python block (line length 100)
+
+
+## Testing and Analysis
+
+Several targets and CMake configuration options are defined as shown below to test the qe-compiler. The standard test suite consists of a format check to ensure the source follows the configured style defined in `.clang-format`, a run of `clang-tidy` to perform code analysis, and a run of the test suite. It is also possible to run additional analysis with Valgrind and the Google Sanitizers below. CMake commands should always be run from a build directory.
+
+Our test suite contains both unit tests (using googletest) and integration tests that exercise the full qss-compiler or qss-opt tools (using LLVM lit).
```
Comment (blakejohnson) [inline CONTRIBUTING.md:269]:
```suggestion
Our test suite contains both unit tests (using googletest) and integration tests that exercise the full qe-compiler or qss-opt tools (using LLVM lit).
```

Comment (blakejohnson) [reply CONTRIBUTING.md:269]: What is `qss-opt`?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+For example using `export LIT_FILTER=Conversion/CodeGen/pulse-gen-drive-0.mlir` will limit the lit
+test to that single test. Note to use this with the `check-tests` target requires passing
+`-DLIT_TEST_EXTRA_ARGS=--allow-empty-runs` to cmake otherwise the testing will stop on the first
+target that has an empty run.
+
+### Setting Paths for Manual Test Runs
+
+The compiler may require static resources at runtime (e.g., target-specific tools or libraries). It will search for the resources in the path defined by the environment variable `QSSC_RESOURCES` or in the designated install location if that variable is not present.
+
+For running `qss-compiler` or `qss-opt` from the build directory, you may use a generated shell script that takes care of setting up the required environment variables as follows:
```
Comment (blakejohnson) [inline CONTRIBUTING.md:295]: I suppose this instance of `qss-compiler` has to stay?

File: README.md [hunk 5]
```diff
@@ -37,10 +50,35 @@ installing with conan. To disable add the option `conan build .. -o pythonlib=Fa
 #### Windows
 Building and running the compiler is supported with [WSL](https://learn.microsoft.com/en-us/windows/wsl/install).
 
+### Static Code Checks
+The easiest, fastest, and most automated way to integrate the formatting into your workflow
+is via [pre-commit](https://pre-commit.com). Note that this tool requires and internet connection
```
Comment (blakejohnson) [inline README.md:55]:
```suggestion
is via [pre-commit](https://pre-commit.com). Note that this tool requires an internet connection
```

File: conan/clang-tools-extra/conanfile.py [hunk 6]
```diff
@@ -48,16 +43,15 @@ class ClangToolsExtraConan(ConanFile):
 
     def source(self):
         git_cache = os.environ.get("CONAN_LLVM_GIT_CACHE")
-        cache_hit = lambda: os.path.exists(f"{git_cache}/.git")
+        cache_hit = os.path.exists(f"{git_cache}/.git")
         cache_arg = f" --reference-if-able '{git_cache}' " if git_cache else ""
 
         if git_cache and cache_hit():
```
Comment (blakejohnson) [inline conan/clang-tools-extra/conanfile.py:49]: You converted `cache_hit` from a function to a boolean. So, I think:
```suggestion
        if git_cache and cache_hit:
```

File: conan/clang-tools-extra/conanfile.py [hunk 7]
```diff
@@ -48,16 +43,15 @@ class ClangToolsExtraConan(ConanFile):
...
-            )
+            self.output.info(f"Cache hit! Some Git objects will be loaded from '{git_cache}'.")
 
         self.run(
-            f"git clone {cache_arg} -b {LLVM_TAG} --single-branch https://github.com/llvm/llvm-project.git"
+            f"git clone {cache_arg} -b {LLVM_TAG} "
+            "--single-branch https://github.com/llvm/llvm-project.git"
         )
 
         if git_cache and not cache_hit():
```
Comment (blakejohnson) [inline conan/clang-tools-extra/conanfile.py:57]: Another instance where you now need a non-function call. (I haven't scanned the file for other places needing the same change):
```suggestion
        if git_cache and not cache_hit:
```

File: conan/llvm/conanfile.py [hunk 8]
```diff
@@ -91,16 +90,15 @@ class LLVMConan(ConanFile):
 
     def source(self):
         git_cache = os.environ.get("CONAN_LLVM_GIT_CACHE")
-        cache_hit = lambda: os.path.exists(f"{git_cache}/.git")
+        cache_hit = os.path.exists(f"{git_cache}/.git")
```
Comment (blakejohnson) [inline conan/llvm/conanfile.py:93]: If you pulled this out above (L22-25), then why keep the redefinition here?

**Review by blakejohnson — CHANGES_REQUESTED** [review]
Other than a few typos, I think you just need to adjust your logic around the changes to `cache_hit` in the conanfile.py files.

File: CONTRIBUTING.md (hunk 4 above)
Comment (taalexander) [reply CONTRIBUTING.md:295]: Yah, I tried to keep any tool mentions consistent for now.

File: CONTRIBUTING.md (hunk 3 above)
Comment (taalexander) [reply CONTRIBUTING.md:269]: I kept this the same as the binaries are still named qss-compiler.

File: conan/llvm/conanfile.py (hunk 8 above)
Comment (taalexander) [reply conan/llvm/conanfile.py:93]: Was a leftover attempt at a flake fix. I've removed.

File: conan/clang-tools-extra/conanfile.py (hunk 7 above)
Comment (taalexander) [reply conan/clang-tools-extra/conanfile.py]: Fixed in dc60201d2a76.

File: conan/clang-tools-extra/conanfile.py (hunk 6 above)
Comment (taalexander) [reply conan/clang-tools-extra/conanfile.py]: Fixed in dc60201d2a76.

File: README.md (hunk 5 above)
Comment (taalexander) [reply README.md]: Fixed in dc60201d2a76.

**Review by blakejohnson — DISMISSED** [review]

**Review by blakejohnson — APPROVED** [review]

**Comment by taalexander** [comment]
This is building now. As it touches the conan files it will clear the cache and take about three hours.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -20,19 +20,14 @@ jobs:
...
           sudo rm -rf "$AGENT_TOOLSDIRECTORY"
           df -h
       - uses: actions/checkout@v3
         with:
           fetch-depth: 0
-      # -- REMOVE ONCE OPEN SOURCE
-      # Needed until Qiskit/qss-qasm is public.
-      - uses: webfactory/ssh-agent@v0.7.0
-        with:
-          ssh-private-key: ${{ secrets.SSH_PRIVATE_KEY }}
```
Comment (blakejohnson) [inline .github/workflows/ci.yml:35]: You may need to restore this section if you want to git clone dependencies via SSH.

Comment (blakejohnson) [reply .github/workflows/ci.yml:35]: Actually, it would be easier just to clone `qe-qasm` via HTTPS.

File: conan/qasm/conanfile.py [hunk 10]
```diff
@@ -35,11 +34,7 @@ class QasmConan(ConanFile):
     description = "A flex/bison parser for OpenQASM v3. A part of the Quantum Engine project."
 
     def source(self):
-        token = os.environ.get("GITHUB_PAT")
-        if token is not None:
-            self.run(f"git clone https://{token}@github.com/openqasm/qe-qasm.git .")
-        else:
-            self.run("git clone git@github.com:openqasm/qe-qasm.git .")
+        self.run("git clone git@github.com:openqasm/qe-qasm.git .")
```
Comment (blakejohnson) [inline conan/qasm/conanfile.py:37]: I believe this is going to fail again unless you change it to:
```suggestion
        self.run("git clone https://github.com/openqasm/qe-qasm.git .")
```

Comment (taalexander) [reply conan/qasm/conanfile.py]: Fixed in [d8539bf](https://github.com/openqasm/qe-compiler/pull/215/commits/d8539bf1267279a3efbcc656b010b21b7ed4ffa7)

File: .github/workflows/clang-tidy-review-comments.yml [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+# that they have been altered from the originals.
+
+# Workaround CI permissions for forks
+# https://github.com/ZedThree/clang-tidy-review?tab=readme-ov-file#usage-in-fork-environments-split-workflow
+name: Post clang-tidy review comments
+
+on:
+  workflow_run:
+    # The name field of the lint action
+    # Defined in ci.yml workflow
```
Comment (blakejohnson) [inline .github/workflows/clang-tidy-review-comments.yml:20]:
```suggestion
    # Defined in lint.yml workflow
```

Comment (taalexander) [reply .github/workflows/clang-tidy-review-comments.yml]: Fixed in f59ebf6

**Review by blakejohnson — APPROVED** [review]

**Pull request merged**